Local monitoring setup with OTEL and Grafana LGTM-stack

martinothamar opened the pull request:
## Description
This PR includes a separate Docker compose file to setup OTEL collector and the Grafana LGTM stack for local monitoring
Includes

* Backends for logs, traces and metrics
* Correlation is built into OTEL, navigation between logs and traces in Grafana UI is configured
* Dashboards published by .NET team for .NET 8/Aspire is automatically provisioned
* Custom dashboard for Altinn apps

Altinn app dashboard:
![image](https://github.com/Altinn/app-localtest/assets/5425986/d2d931e3-5f49-4a68-8661-f8a1dc261581)
![image](https://github.com/Altinn/app-localtest/assets/5425986/f3969352-4081-4de2-8621-68c1e07b09ed)


ASP.NET core dashboard:
![image](https://github.com/Altinn/app-localtest/assets/5425986/73182a66-7450-418c-b764-644a7080d4c3)

Explore logs:
![image](https://github.com/Altinn/app-localtest/assets/5425986/7e13dafe-93f9-47de-a080-bb62abb4d853)

Explore traces:
![image](https://github.com/Altinn/app-localtest/assets/5425986/bce6e0ee-a76c-4252-916a-4cd69e497571)

Explore metrics:
![image](https://github.com/Altinn/app-localtest/assets/5425986/48af2893-c92c-4307-a2e5-4ffaa93e0947)

Decisions:
* How "built in" should monitoring stack be - included in the primary docker compose file or a separate one? The backends are relatively lightweight, but not free
* Grafanas own LGTM-stack development image can also be considered but it is far less flexible, and harder to match the fully deployed solution: https://github.com/grafana/docker-otel-lgtm

TODOs
* Instrument app-lib a lot better
* Continuous profiling setup? It is being considered for OTEL AFAIK, might be too early
* Dashboard for runtime metrics (exceptions, threads count, lock contention etc)

## Related Issue(s)
- Issue: https://github.com/Altinn/app-lib-dotnet/issues/587
- Epic: https://github.com/Altinn/app-lib-dotnet/issues/580

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

ivarne commented:
Not a requirement for this PR, but I'm missing a link in the localtest dotnet app to make this more discoverable, and ideally a message informing that local telemetry is not running if it isn't (just send a http request to verify).

Docs does currently not suggest using the `Makefile`, so adding everything together in a single makefile would be nice. Potentially behind a `--profile` flag (as we use in `--profile localtest`).

For the other services we have separate compose files for docker and podman. Would that be required here also?

martinothamar commented:
> but I'm missing a link in the localtest dotnet app to make this more discoverable

Good point, now there is a Grafana menu item 
![image](https://github.com/Altinn/app-localtest/assets/5425986/721a0f80-8475-44d6-bc2b-72f3820495ac)

RE: 
> Docs does currently not suggest using the Makefile, so adding everything together in a single makefile would be nice. Potentially behind a --profile flag (as we use in --profile localtest).

and 

> IMO it should be in a separate one.

Opted for using compose profiles in the latest commit

* `--profile "monitoring"` brings up existing setup + monitoring profile
* `--profile "*"` brings up existing setup + every other profile
* No profile specified brings up the existing setup


> What are we missing exactly from running that image? Is it adding dashboards?

I was guessing that configuration would be more of a hassle but looking at it now maybe it isn't. I won't be able to specify `command` in the compose file, and for dashboards I will have to get the grafana version number, but the rest seems like it should work fine (the actual config files and such). I will give it a shot

ivarne commented on docker-compose.yml:
```diff
@@ -1,5 +1,3 @@
-version: "3.6"
-
```
Has the recommendation to specify version been removed?

martinothamar replied:
Yeah for the latest versions of the CLI there is a message saying it could be removed (it hasn't really been needed for >3 AFAIK)

ivarne commented on docker-compose.yml:
```diff
@@ -77,5 +75,80 @@ services:
+    volumes:
+      - ./infra/otel-collector.yaml:/etc/otel-collector.yaml
+    ports:
+      - "1888:1888"   # pprof extension
+      - "8888:8888"   # Prometheus metrics exposed by the collector
+      - "8889:8889"   # Prometheus exporter metrics
+      - "13133:13133" # health_check extension
+      - "4317:4317"   # OTLP gRPC receiver
+      - "55679:55679" # zpages extension
```
This is a lot of ports to expose externally on the host machine. Are every one required to be accessible from user browser/locally running app? Especially `8888` seems like something that a dev would use for another purpose.

ivarne commented on docker-compose.yml:
```diff
@@ -77,5 +75,80 @@ services:
+      - GF_FEATURE_TOGGLES_ENABLE="traceqlEditor traceToMetrics metricsSummary"
+      - GF_LOG_LEVEL=error
+    ports:
+      - "3000:3000"
```
Maybe consider a more random looking port number in case someone runs anything else on 3000?

ivarne replied:
Or run this through the nginx "load balancer" (proxy), so that it does not need to be on a separate port.

martinothamar replied:
Went with the proxy approach 👍

ivarne reviewed (approved):
I think this looks good enough for merging as a preview, but had a few thoughts while reading through.

martinothamar commented:
Running out of time a bit here so I'll keep the LGTM stack as is for now. Other than that

* Grafana is behind proxy (no exposed port)
* Only OTel GRPC port exposed on OTel collector
* Some updates to the dashboard (will update screenshot in a bit)
* Updated README
* Implemented podman stuff
* Implemented a simpler `run.cmd` cross-platform script and documented it as preview

martinothamar commented on README.md:
````diff
@@ -54,6 +57,14 @@ Start the containers with the following command:
+
+```shell
+podman compose --file podman-compose.yml --profile "monitoring" up -d --build
+# Grafana should be available at http://local.altinn.cloud/grafana
````
Different port here by default

HauklandJ reviewed (approved):
LGTM looks good to me
